rust/kernel/thread.rs

```diff
@@ -0,0 +1,508 @@
+primitive_arg_struct! {u8}
+primitive_arg_struct! {u16}
+primitive_arg_struct! {u32}
+primitive_arg_struct! {u64}
+primitive_arg_struct! {usize}
+
+/// Bridge function from Rust ABI to C.
+extern "C" fn bridge<A, F: ThreadFunc<A>>(data: *mut c_types::c_void) -> i32
+where
+    A: Send,
+{
+    let arg = unsafe { F::Arg::from_raw(data) };
+
+    match F::func(arg) {
+        Ok(_) => 0,
+        Err(e) => e.to_kernel_errno(),
+    }
+}
+
+/// Helper struct for thread that just call its argument as a closure.
+///
+/// We use double boxing in the implementation, because 1) we need to make sure
+/// the thread argument live longer enough for the new thread to use (box #1),
+/// and `FnOnce()` is a fat pointer that doesn't fit in `*mut c_void` (box #2).
+struct ClosureCallFunc;
+
+impl ThreadFunc<Box<Box<dyn FnOnce() -> KernelResult<()> + Send>>> for ClosureCallFunc {
+    type Arg = BoxArg;
+    fn func(arg: Box<Box<dyn FnOnce() -> KernelResult<()> + Send>>) -> KernelResult<()> {
+        // Just calls it.
+        arg()
+    }
+}
+
+/// Function passed to `kthread_create_on_node` as the thread function pointer.
+#[no_mangle]
+unsafe extern "C" fn rust_thread_func(data: *mut c_types::c_void) -> c_types::c_int {
+    // `Box::from_raw()` to get the ownership of the closure.
+    let c = Box::from_raw(data as *mut Box<dyn FnOnce() -> KernelResult<()>>);
+
+    match c() {
+        Ok(_) => 0,
+        Err(e) => e.to_kernel_errno(),
+    }
+}
+
+/// A kernel thread handle.
+pub struct Thread {
+    /// Pointer to the kernel thread.
+    task: *mut bindings::task_struct,
+}
+
+impl Thread {
+    /// Creates a new thread using a C-style function pointer.
+    ///
+    /// No extra memory allocation for thread creation. Use this when closure
+    /// allocation overhead is unacceptable or there is already a C style
+    /// thread function. Otherwise, please consider using [`Thread::try_new`].
+    ///
+    /// # Safety
+    ///
+    /// This function actually doesn't dereference `arg` or call `f`, so even if
+    /// the users pass incorrect parameters this function won't run into
+    /// trouble. But if the users provide incorrect `arg` or `f` the new
+    /// thread will corrupt memory or do other unsafe behaviors, so
+    /// make it `unsafe`. Otherwise, it is [`Thread::wake_up`] that should be
+    /// made as `unsafe` because it will trigger the call of the unsafe function
+    /// `f`, this is undesirable given that [`Thread::wake_up`] on a
+    /// [`Thread::try_new`] created thread is safe.
+    ///
+    /// The safety requirements of calling this function are:
+    ///
+    /// - Make sure `arg` is a proper pointer that points to a valid memory
+    ///   location when the new thread begins to run.
+    ///
+    /// - Make sure `f` is a proper function pointer and `f` handles `arg`
+    ///   correctly.
+    ///
+    /// # Context
+    ///
+    /// This function might sleep due to the memory allocation and waiting for
```

Comment:
This one does not allocate, right?

Reply:
The kernel interface `kthread_create_on_node` **always** allocates._

Reply:
And this comment line basically means since we call a function that will allocate memory, so we cannot call the function in atomic contexts.

Reply:
Ah, yes, I thought the comment was about the data argument passed to the thread (i.e. a copy-paste error from the allocating version with the double box etc.), never mind.

Perhaps add where the allocation happens:

```suggestion
    /// This function might sleep due to the memory allocation in `kthread_create_on_node` and waiting for
```

Reply:
The "waiting for completion" also happens in the `kthread_create_node` ;-)

Reply:
So the whole sentence is "This function might sleep due to A and B in `kthread_create_on_node`", where A is "the memory allocation" and B is "waiting for completion". Maybe I just replace with something like:

/// Because `kthread_create_on_node` will allocate memory and wait for thread creation completed, therefore this funciton might sleep.

Reply:
Perhaps:

    This function might sleep in `kthread_create_on_node` due to A and B .....

Reply:
Ah, I see the confusion: I was in the Conversation tab which didn't show the next lines. Now I see you already had the note on `kthread_create_on_node`. Never mind, it is fine like that, but if you want to move it to make it clearer, feel free.